webapi/src/main/scala/org/knora/webapi/slice/admin/domain/service/ProjectExportService.scala

```diff
@@ -0,0 +1,209 @@
+        outputStream = maybeBufferedFileOutputStream.get,
+        rdfFormat = TriG
+      )
+
+      val combiningRdfProcessor = new CombiningRdfProcessor(formattingStreamProcessor)
+      formattingStreamProcessor.start()
+
+      for (namedGraphTrigFile: NamedGraphTrigFile <- namedGraphTrigFiles) {
+        val namedGraphTry: Try[Unit] = Try {
+          rdfFormatUtil.parseWithStreamProcessor(
+            rdfSource =
+              RdfInputStreamSource(new BufferedInputStream(Files.newInputStream(namedGraphTrigFile.dataFile))),
+            rdfFormat = TriG,
+            rdfStreamProcessor = combiningRdfProcessor
+          )
+        }
+
+        Files.delete(namedGraphTrigFile.dataFile)
+
+        namedGraphTry match {
+          case Success(_)  => ()
+          case Failure(ex) => throw ex
+        }
+      }
+
+      formattingStreamProcessor.finish()
+    }
+
+    maybeBufferedFileOutputStream.foreach(_.close)
+
+    trigFileTry match {
+      case Success(_)  => ()
+      case Failure(ex) => throw ex
+    }
+  }
+
+  override def exportProjectTriples(project: KnoraProject, file: Path): Task[Path] = {
+    val tempDir = Files.createTempDirectory(project.shortname)
+    for {
+      _                 <- ZIO.logInfo("Downloading project data to temporary directory " + tempDir.toAbsolutePath)
+      projectOntologies <- ontologyRepo.findByProject(project).map(_.map(_.ontologyMetadata.ontologyIri.toIri))
+
+      // Download the project's named graphs.
+
+      projectDataNamedGraph: IRI = ProjectADMService.projectDataNamedGraphV2(project).value
+      graphsToDownload: Seq[IRI] = projectOntologies :+ projectDataNamedGraph
+      projectSpecificNamedGraphTrigFiles: Seq[NamedGraphTrigFile] =
+        graphsToDownload.map(graphIri => NamedGraphTrigFile(graphIri = graphIri, tempDir = tempDir))
+
+      _ <- ZIO.foreachDiscard(projectSpecificNamedGraphTrigFiles) { trigFile =>
+             for {
+               fileWrittenResponse <- messageRelay.ask[FileWrittenResponse](
+                                        NamedGraphFileRequest(
+                                          graphIri = trigFile.graphIri,
+                                          outputFile = trigFile.dataFile,
+                                          outputFormat = TriG
+                                        )
+                                      )
+             } yield fileWrittenResponse
+           }
+
+      // Download the project's admin data.
+
+      adminDataNamedGraphTrigFile = NamedGraphTrigFile(graphIri = adminDataGraph, tempDir = tempDir)
+
+      projectIri      = project.id.value
+      adminDataSparql = twirl.queries.sparql.admin.txt.getProjectAdminData(projectIri)
+      _ <- triplestoreService.sparqlHttpConstructFile(
+             sparql = adminDataSparql.toString(),
+             graphIri = adminDataNamedGraphTrigFile.graphIri,
+             outputFile = adminDataNamedGraphTrigFile.dataFile,
+             outputFormat = TriG
+           )
```

Comment:
these seem like a lower level of abstraction. Could it be moved to some service/repo?

Reply:
In my opinion, this is the knowledge that belongs exactly in this service. It must own the query and the knowledge where safe the file.